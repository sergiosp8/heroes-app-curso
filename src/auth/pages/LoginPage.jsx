import { useContext } from 'react'
import { useNavigate } from 'react-router-dom'
import { AuthContext } from '../context/AuthContext'

export const LoginPage = () => {
  const navigate = useNavigate()
  const { onLogin } = useContext(AuthContext)

  const handleLogin = () => {
    const lastPath = localStorage.getItem('lastPath') || '/'

    onLogin('Sergio Sosa')
    navigate(lastPath, { replace: true })
  }

  return (
    <div className="container mt-5">
      <h1>Login</h1>
      <hr />
      <button className="btn btn-primary" onClick={handleLogin}>
        Login
      </button>
    </div>
  )
}
