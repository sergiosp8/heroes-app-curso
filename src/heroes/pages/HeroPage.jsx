import { useMemo } from 'react'
import { Navigate, useNavigate, useParams } from 'react-router-dom'
import { getHeroById } from '../helpers'

export const HeroPage = () => {
  const { id } = useParams()
  const navigate = useNavigate()

  const hero = useMemo(() => getHeroById(id), [id])

  const onNavegateBack = () => {
    navigate(-1)
  }

  if (!hero) {
    return <Navigate to="/marvel" />
  }

  return (
    <div className="row mt-5 animate__animated animate__fadeInLeft">
      <div className="col-4">
        <img src={`/assets/heroes/${id}.jpg`} className="img-thumbnail " alt={hero.superhero} />
      </div>
      <div className="col-8">
        <h3>{hero.superhero}</h3>
        <ul className="list-group lis-group-flush">
          <li className="list-grou-item">
            <b>Alter ego :</b> {hero.alter_ego}
          </li>
          <li className="list-grou-item">
            <b>Publisher :</b> {hero.publisher}
          </li>
          <li className="list-grou-item">
            <b>First appearence :</b> {hero.first_appearance}
          </li>
        </ul>
        <h5 className="mt-5">Characters</h5>
        <p>{hero.characters}</p>
        <button className="btn btn-primary" onClick={onNavegateBack}>
          Back
        </button>
      </div>
    </div>
  )
}
