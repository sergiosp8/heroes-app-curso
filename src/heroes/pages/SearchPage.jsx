import { useForm } from '../../hooks/useForm'
import { useLocation, useNavigate } from 'react-router-dom'

import { getHeroesByName } from '../helpers'
import { HeroCard } from '../components'
import queryString from 'query-string'

export const SearchPage = () => {
  const navigate = useNavigate()
  const location = useLocation()
  const { q = '' } = queryString.parse(location.search)
  const { searchText, onInputChange } = useForm({ searchText: q })
  const heroes = getHeroesByName(q)

  const showSearch = q.length === 0
  const showError = q.length > 0 && heroes.length === 0

  const onSearchSubmit = (e) => {
    e.preventDefault()
    navigate(`?q=${searchText.toLowerCase().trim()}`)
  }

  return (
    <>
      <h1>Search</h1>
      <hr />

      <div className="row">
        <div className="col-5">
          <h4>Searching...</h4>
          <hr />
          <form aria-label="form" onSubmit={onSearchSubmit}>
            <input
              type="text"
              placeholder="Search a horo"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={onInputChange}
            />
            <button className="mt-1 btn btn-primary">Search</button>
          </form>
        </div>
        <div className="col-7">
          <h4>Result</h4>
          <hr />

          {/* {q === '' ? (
            <div className="alert alert-primary">Search a Hero</div>
          ) : (
            heroes.length === 0 && (
              <div className="alert alert-danger">
                No hero with <b>{q}</b>
              </div>
            )
          )} */}
          <div
            className="alert alert-primary animate__animated animate__fadeIn"
            style={{ display: showSearch ? '' : 'none' }}
          >
            Search a Hero
          </div>

          <div
            className="alert alert-danger animate__animated animate__fadeIn"
            aria-label="danger"
            style={{ display: showError ? '' : 'none' }}
          >
            No hero with <b>{q}</b>
          </div>

          <div>
            {heroes.map((hero) => (
              <HeroCard key={hero.id} {...hero} />
            ))}
          </div>
        </div>
      </div>
    </>
  )
}
