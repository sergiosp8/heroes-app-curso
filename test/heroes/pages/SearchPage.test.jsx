import { fireEvent, render, screen } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import { SearchPage } from '../../../src/heroes/pages/SearchPage'

const mockedUsedNavigate = jest.fn()
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate
}))

describe('Pruebas en Search Screen', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })
  test('debe de mostrarse con valores por defecto', () => {
    const { container } = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    )
    expect(container).toMatchSnapshot()
    // screen.debug()
  })

  test('debe de mostrar a batman y el input con el valor del queryString', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchPage />
      </MemoryRouter>
    )

    const input = screen.getByRole('textbox')
    expect(input.value).toBe('batman')

    const img = screen.getByRole('img')
    expect(img.src).toContain('/assets/heroes/dc-batman.jpg')

    const alert = screen.getByLabelText('danger')
    expect(alert.style.display).toBe('none')

    // screen.debug()
  })

  test('debe de mostrar el div danger cuando no hay resultados', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=batman123']}>
        <SearchPage />
      </MemoryRouter>
    )
    const alert = screen.getByLabelText('danger')
    expect(alert.style.display).toBe('')
    // screen.debug()
  })

  test('debe de llamar el navigate a la pantalla nueva', () => {
    render(
      <MemoryRouter initialEntries={['/search']}>
        <SearchPage />
      </MemoryRouter>
    )

    const input = screen.getByRole('textbox')
    fireEvent.change(input, { target: { value: 'superman' } })
    const form = screen.getByLabelText('form')
    fireEvent.submit(form)

    expect(mockedUsedNavigate).toHaveBeenCalledWith('?q=superman')
    // screen.debug()
  })
})
