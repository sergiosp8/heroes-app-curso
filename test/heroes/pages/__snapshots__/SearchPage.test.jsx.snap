// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en Search Screen debe de mostrarse con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching...
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a horo"
          type="text"
          value=""
        />
        <button
          class="mt-1 btn btn-primary"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Result
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search a Hero
      </div>
      <div
        aria-label="danger"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
      <div />
    </div>
  </div>
</div>
`;
